kernel/core/extend/pagemem.c

```diff
@@ -17,33 +17,37 @@ void init_pgd()
 
 	/*************** Mapping addresses and setting up PGD and PTBs ***************/
 	// Mapping kernel memory
-	krn_identity_map(&PGD[0], &PTB[0], KERNEL_START, KERNEL_END);
+	krn_identity_map(&PGD[0], KERNEL_START, KERNEL_END);
 
 	// Mapping kernel stack
-	krn_identity_map(&PGD[0], &PTB[0], KSTACK_START, KSTACK_END);
+	krn_identity_map(&PGD[0], KSTACK_START, KSTACK_END);
 
 	// Mapping GDT
-	krn_identity_map(&PGD[0], &PTB[0], GDT_START, GDT_END);
+	krn_identity_map(&PGD[0], GDT_START, GDT_END);
 
 	// Mapping TSS
-	krn_identity_map(&PGD[0], &PTB[0], TSS_START, TSS_END);
+	krn_identity_map(&PGD[0], TSS_START, TSS_END);
 
-	// Mapping memory for both PTBs
-	krn_identity_map(&PGD[0], &PTB[0], (offset_t)&PTB[0], (offset_t)&PTB[0] + 2*PTB_SIZE - 1);
+	// Mapping memory for PGDs
+	krn_identity_map(&PGD[0], PGD_START, PGD_START + NUM_PGD*PGD_SIZE - 1);
+
+	// Mapping memory for PTBs
+	krn_identity_map(&PGD[0], PTB_START, PTB_START + NUM_PTB*PTB_SIZE - 1);
 
 	// Mapping memory where PGD is located so it can be accessed
-	krn_forced_map(&PGD[0], &PTB[0], 0xc0000000, (offset_t)&PGD[0], PTE_OFFSET);
+	// krn_forced_map(&PGD[0], 0xc0000000, &PGD[0], PGD_SIZE);
 
 	/*************** For testing only: map for user for testing ring 3 jump ***************/
 	/*********************  TODO: Should ultimately be kernel mapped  *********************/
 	// Mapping user memory
-	usr_identity_map(&PGD[0], &PTB[0], USER_START, USER_END); // USR
+	usr_identity_map(&PGD[0], USER_START, USER_END); // USR
 
 	// Mapping userland, for now userland is in Kernel memory
-	usr_identity_map(&PGD[0], &PTB[0], (offset_t)userland, (offset_t)userland + PDE_OFFSET - 1);
-	usr_identity_map(&PGD[0], &PTB[0], (offset_t)printf, (offset_t)userland + PDE_OFFSET - 1);
+	usr_identity_map(&PGD[0], (offset_t)userland, (offset_t)userland + PTE_OFFSET - 1);
+	usr_identity_map(&PGD[0], (offset_t)printf, (offset_t)userland + PTE_OFFSET - 1);
```

Comment:
Yes, we need to find a way to put user functions somewhere... copy? Must be next feature!

Reply:
By defining a .user section in linker script and adding a section attribute to the functions!